Use new MMS APIs available from Lollipop and Lollipop MR1

mcginty opened the pull request:
Fixes #1937

mcginty commented:
Diff updated, tested MMS to/from lollipop and non-lollipop device with good results :).

moxie0 commented on src/org/thoughtcrime/securesms/ConversationItem.java:
```diff
@@ -314,8 +314,10 @@ private void setSentStatusIcons() {
+    if (indicatorText != null) {
+      indicatorText.setText(R.string.ConversationItem_click_for_details);
+      indicatorText.setVisibility(View.VISIBLE);
+    }
```
why is this in this diff?

mcginty replied:
i can throw it into another PR, but causes NPEs if there's an incoming failure

moxie0 replied:
is the npe specific to this change?

mcginty replied:
No, just came across it during testing. Busting it out into another PR.

moxie0 commented on src/org/thoughtcrime/securesms/mms/IncomingLollipopMmsConnection.java:
```diff
@@ -0,0 +1,105 @@
+    }
+    Log.w(TAG, "code: " + getResultCode() + ", result string: " + getResultData());
+
+    finished = true;
```
this has to be in the synchronized block

moxie0 commented on src/org/thoughtcrime/securesms/mms/OutgoingLollipopMmsConnection.java:
```diff
@@ -0,0 +1,101 @@
+  private long messageId;
+  private byte[] pduBytes;
+
+  public OutgoingLollipopMmsConnection(Context context, byte[] pduBytes, long messageId) {
```
pdubytes should be a paramter of send, and i don't think this class should know anything about message ids

moxie0 commented on src/org/thoughtcrime/securesms/mms/OutgoingLegacyMmsConnection.java:
```diff
@@ -0,0 +1,164 @@
+
+  private byte[] mms;
+
+  public OutgoingLegacyMmsConnection(Context context, byte[] mms) throws ApnUnavailableException {
```
mms should be a parameter of send

moxie0 commented on src/org/thoughtcrime/securesms/mms/IncomingLollipopMmsConnection.java:
```diff
@@ -0,0 +1,105 @@
+        while (!finished) Util.wait(this, 30000);
+      }
+
+      context.getApplicationContext().unregisterReceiver(this);
```
should probably be finally

moxie0 commented on src/org/thoughtcrime/securesms/mms/IncomingLollipopMmsConnection.java:
```diff
@@ -0,0 +1,105 @@
+  private long    messageId;
+  private boolean finished;
+
+  public IncomingLollipopMmsConnection(Context context, String contentLocation, long messageId) {
```
contentlocation should be a parameter of receive, and i don't think this class should know anything about message ids

moxie0 commented on src/org/thoughtcrime/securesms/mms/IncomingLegacyMmsConnection.java:
```diff
@@ -0,0 +1,148 @@
+                                                       transactionId,
+                                                       PduHeaders.STATUS_RETRIEVED);
+
+      OutgoingLegacyMmsConnection connection = new OutgoingLegacyMmsConnection(context, new PduComposer(context, notifyResponse).make());
```
these classes should probably be peers, not dependencies

moxie0 commented on src/org/thoughtcrime/securesms/ConversationActivity.java:
```diff
@@ -666,7 +666,7 @@ private void initializeMmsEnabledCheck() {
       @Override
       protected Boolean doInBackground(Void... params) {
-        return OutgoingMmsConnection.isConnectionPossible(ConversationActivity.this);
+        return OutgoingLegacyMmsConnection.isConnectionPossible(ConversationActivity.this);
```
shouldn't there be a lollipop check somewhere around here now?

mcginty replied:
that's what I forgot to do! Yeah, I was looking into how to do that check on Lollipop and obviously got distracted. I added a util method to check for both Lollipop and legacy mms capabilities. Right now we don't have any code in to override the APN settings in Lollipop, so I should disable the settings or add in overrides in this PR too.

moxie0 commented on src/org/thoughtcrime/securesms/util/Util.java:
```diff
@@ -279,4 +284,11 @@ public static SecureRandom getSecureRandom() {
+  public static boolean isMmsCapable(Context context) {
+    return (VERSION.SDK_INT >= VERSION_CODES.LOLLIPOP)
+           ? SmsManager.getDefault().getCarrierConfigValues().getBoolean(SmsManager.MMS_CONFIG_MMS_ENABLED)
+           : OutgoingLegacyMmsConnection.isConnectionPossible(context);
```
I think this method is written too generically. IIRC, what we really want to know is whether we need to prompt the user to enter their custom MMSC information into TextSecure in the case where we can't access it for some reason.  For lollipop, that should never be true, since we're delegating everything to the system.

mcginty replied:
I'm modifying that as we speak :)

mcginty commented:
Now displays error if MMS isn't enabled by carrier in lollipop and higher, otherwise will go the normal manual MMS route.

Disabled manual MMS controls on lollipop as well since they're ignored. A TODO could be to integrate the new SocketFactory API in the legacy sending logic to support manual overrides, but that's for another PR.

moxie0 commented on src/org/thoughtcrime/securesms/mms/IncomingLegacyMmsConnection.java:
```diff
@@ -0,0 +1,146 @@
+                                                       transactionId,
+                                                       PduHeaders.STATUS_RETRIEVED);
+
+      OutgoingLegacyMmsConnection connection = new OutgoingLegacyMmsConnection(context);
```
I still think these classes should be peers and not dependencies

mcginty replied:
And have another wrapper class that implements the OutgoingMmsConnection interface and uses both Incoming/OutgoingLegacyMmsConnection?

moxie0 replied:
Some might say MmsDownloadJob is that interface. You might even call it a "controller." =)

mcginty replied:
The outgoing lollipop API appears to send the confirmation automatically, so I felt it was most symmetric/nice to have our legacy implementation do the same thing.

moxie0 replied:
I think it's structurally weird, and could result in some functional weirdness (what happens with a failure on just the ack part?), but I can live if it's what you want to do.

mcginty replied:
I think the Lollipop API was kind of poorly designed for that reason (if your receiver fails for some reason, you'd like to be able to fetch it again).

moxie0 commented on src/org/thoughtcrime/securesms/ConversationActivity.java:
```diff
@@ -540,11 +541,19 @@ private void handleAddAttachment() {
+                                      .content(R.string.ConversationActivity_mms_not_supported_message)
+                                      .iconRes(R.drawable.ic_error_red_24dp)
+                                      .neutralText(android.R.string.ok)
+                                      .show();
```
Should we even do this?  We could just have the message fail as usual.

Maybe it's nice, but it could also cause a wrinkle in the "auto-upgrade" path when sending to recipients who are TS users but aren't yet in the synchronized directory.

mcginty replied:
you're totally right. letting it fail.

moxie0 commented on src/org/thoughtcrime/securesms/jobs/MmsDownloadJob.java:
```diff
@@ -86,71 +101,32 @@ public void onRun(MasterSecret masterSecret)  {
-    byte[]   transactionId   = notification.get().getTransactionId();
-    MmsRadio radio           = MmsRadio.getInstance(context);
+    String contentLocation = new String(notification.get().getContentLocation());
+    byte[] transactionId = notification.get().getTransactionId();
```
alignment

moxie0 commented on src/org/thoughtcrime/securesms/jobs/MmsDownloadJob.java:
```diff
@@ -86,71 +101,32 @@ public void onRun(MasterSecret masterSecret)  {
+    download(masterSecret, contentLocation, transactionId);
+  }
 
+  private void download(MasterSecret masterSecret, String contentLocation, byte[] transactionId) {
```
I think it makes sense to merge these methods now.

moxie0 commented on src/org/thoughtcrime/securesms/jobs/MmsSendJob.java:
```diff
@@ -90,59 +89,36 @@ public void onCanceled() {
   }
 
-  public MmsSendResult deliver(MasterSecret masterSecret, SendReq message)
+  private MmsSendResult deliver(MasterSecret masterSecret, SendReq message)
```
I don't think this needs to be a separate method anymore.

moxie0 commented on src/org/thoughtcrime/securesms/mms/IncomingLegacyMmsConnection.java:
```diff
@@ -0,0 +1,146 @@
+
+  private byte[] transactionId;
+
+  public IncomingLegacyMmsConnection(Context context, Apn apn, byte[] transactionId) {
```
i still think the latter parameters should be params of retrieve

mcginty replied:
lol. just looked at how these notifications are sent back if we don't pass the transactionId to the lollipop API. It's not even serialized by the PduComposer. I'm not sure what it's there for.

moxie0 commented on src/org/thoughtcrime/securesms/mms/IncomingLollipopMmsConnection.java:
```diff
@@ -0,0 +1,100 @@
+  private String  contentLocation;
+  private boolean finished;
+
+  public IncomingLollipopMmsConnection(Context context, String contentLocation) {
```
latter param should be of retrieve

moxie0 commented on src/org/thoughtcrime/securesms/ConversationActivity.java:
```diff
@@ -540,11 +541,13 @@ private void handleAddAttachment() {
 
   private void handleManualMmsRequired() {
-    Toast.makeText(this, R.string.MmsDownloader_error_reading_mms_settings, Toast.LENGTH_LONG).show();
+    if (Util.canOverrideMms()) {
```
Is this check necessary?

moxie0 commented on src/org/thoughtcrime/securesms/mms/OutgoingLollipopMmsConnection.java:
```diff
@@ -0,0 +1,94 @@
+      SmsManager.getDefault().sendMultimediaMessage(context, contentUri, null, null,
+                                                    PendingIntent.getBroadcast(context, 1, new Intent(ACTION), PendingIntent.FLAG_ONE_SHOT));
+
+      while (!finished) Util.wait(this, 30000);
```
Why does this have a timeout if you loop forever?

mcginty replied:
yeah, that logic doesn't make sense (I had the timeout before I did the looped conditional). I changed it so we actually timeout after 30s of waiting.

moxie0 commented on src/org/thoughtcrime/securesms/mms/IncomingLollipopMmsConnection.java:
```diff
@@ -0,0 +1,105 @@
+        if (System.currentTimeMillis() >= timeoutExpiration) {
+          throw new TimeoutException("timeout when waiting for MMS");
+        }
+      }
```
You've written this so that it's only possible for a single thread to use this at a time, and such that an instance can only be used once before it needs to be destroyed.  Personally I don't think that's great, but given that it's the case, this logic doesn't really make sense.  You call notify in exactly one place, immediately after setting finished to true.  Just do if(!finished) Util.wait(this, 30000)

If you want this to be thread-safe and reusable, that's fine with me too, but I'd like you to pick a paradigm.

mcginty replied:
Java's own documentation instructs you to always wrap an Object.wait() with a conditional, since "interrupts and spurious wakeups are possible." This logic's not about thread safety, it's about implementing waits as correctly as possible.

moxie0 replied:
Spurious wakeups should only happen if the process receives a signal.  I guess it is entirely possible that Android uses signals in some way I don't know about.

mcginty replied:
I don't know either, but figured better safe to do it this way. Let me know if you want to just go with the simple version instead.

moxie0 commented on src/org/thoughtcrime/securesms/mms/IncomingLollipopMmsConnection.java:
```diff
@@ -0,0 +1,105 @@
+  public static final String ACTION = IncomingLollipopMmsConnection.class.getCanonicalName() + "MMS_DOWNLOADED_ACTION";
+  private static final String TAG = IncomingLollipopMmsConnection.class.getSimpleName();
+
+  private Context context;
```
final

moxie0 commented on src/org/thoughtcrime/securesms/mms/IncomingLollipopMmsConnection.java:
```diff
@@ -0,0 +1,105 @@
+  @TargetApi(VERSION_CODES.LOLLIPOP)
+  public synchronized RetrieveConf retrieve(String contentLocation, byte[] transactionId) throws MmsException {
+    context.getApplicationContext().registerReceiver(this, new IntentFilter(ACTION));
+    long nonce = System.currentTimeMillis();
```
inlign

moxie0 commented on src/org/thoughtcrime/securesms/mms/IncomingLollipopMmsConnection.java:
```diff
@@ -0,0 +1,105 @@
+    long nonce = System.currentTimeMillis();
+    try {
+      PendingIntent pendingIntent = PendingIntent.getBroadcast(context, 1, new Intent(ACTION), PendingIntent.FLAG_ONE_SHOT);
+      Uri contentUri = ContentUris.withAppendedId(MmsBodyProvider.CONTENT_URI, nonce);
```
align

moxie0 commented on src/org/thoughtcrime/securesms/mms/OutgoingLollipopMmsConnection.java:
```diff
@@ -0,0 +1,103 @@
+  @TargetApi(VERSION_CODES.LOLLIPOP)
+  public synchronized SendConf send(byte[] pduBytes) throws UndeliverableMessageException {
+    context.getApplicationContext().registerReceiver(this, new IntentFilter(ACTION));
+    long nonce = System.currentTimeMillis();
```
inlign

moxie0 commented on src/org/thoughtcrime/securesms/mms/OutgoingLollipopMmsConnection.java:
```diff
@@ -0,0 +1,103 @@
+        if (System.currentTimeMillis() >= timeoutExpiration) {
+          throw new TimeoutException("timeout when waiting for MMS");
+        }
+      }
```
this looks familiar...

mcginty commented:
Addressed comments. Decided to create an abstract LollipopMmsConnection class that manages most of the synchronization logic and wraps it in the form of a transaction. This way the class doesn't render itself totally useless after the first transaction.

Also decided to abstract out all the content provider stuff and just have a MMS body "pointer" for temporary receptacles needed to transact with the Android API.

moxie0 commented on src/org/thoughtcrime/securesms/jobs/MmsDownloadJob.java:
```diff
@@ -222,26 +164,8 @@ private void storeRetrievedMms(MasterSecret masterSecret, String contentLocation
-
-  private void handleDownloadError(MasterSecret masterSecret, long messageId, long threadId,
-                                   int downloadStatus, String error, boolean automatic)
+  private static void handleDownloadError(Context context, MasterSecret masterSecret, long messageId, long threadId,
```
static?
